### PR TITLE
Layout + template switching mechanism

### PR DESCRIPTION
**Pull request by pkarw**
### Related issues

#1713, #1688

### Short description and why it's useful

I've changed the way SSR pages are generated:
- now distinct routes can set "context.serverOutputTemplate" in `asyncData` and by doing so You can skip using `dist/index.html` (which contains typical HTML5 elements - like `<head` ...). This is important when we're going to generate AMPHTML pages (that can not contain any `<script>` tags) - either xml files - You name it
- distinct routes can set `meta.layout` and by doing so switch the previously constant `App.vue` layout file - this is what @mercs600 proposed some time ago in slightly changed form

### Upgrade Notes and Changelog

The `themes/default/index.template.html` and `src/index.template.html` template files have been modified by adding `renderStyles()`, `renderState()` helpers. AFAIK nobody is modifying these templates - but if sbd. did - he/she must change it accordingly otherwise styles won't be rendered properly

- [x] No upgrade steps required (100% backward compatibility)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/doc/Upgrade%20notes.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

### Contribution and curently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and [refactoring plan for them](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/refactoring-to-modules.md)

### CODE REVIEW

File: src/extensions/raw-output-example/pages/RawOutputExample.vue [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  <raw-content>
+    This page is using empty layout set in routes + no html template
+  </raw-content>
+</template>
+
+<script>
+export default {
+  name: 'RawOutputExample',
+  asyncData ({ store, route, context }) {
+    context.response.setHeader('Content-Type', 'text/xml')
```
Comment (pkarw) [inline src/extensions/raw-output-example/pages/RawOutputExample.vue]: Here You can now set from the Vue component level server response headers + choose template

File: src/extensions/raw-output-example/router.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import RawOutputExample from './pages/RawOutputExample.vue'
+
+export default [
+  { path: '/raw-output-example.xml', component: RawOutputExample, serverOutputTemplate: '', meta: { layout: 'empty' } }
```
Comment (pkarw) [inline src/extensions/raw-output-example/router.js]: You can now also set the layout from the routes level

**Comment by filrak** [comment]
Just one question regarding this one
"distinct routes can set meta.layout and by doing so switch the previously constant App.vue layout file - this is what @mercs600 proposed some time ago in slightly changed form"

Shouldn't we do it in a vue way just by creating one component inside `App.vue` that is switchable. I think the proboem solved by this solution is default-theme specific. This problem will no longer occur on new default theme since i wanted to make layouts system there (like in nuxt.js, they had the same problem and made it in vue way).

**Comment by filrak** [comment]
Also, can we add posibility to server side render onlyc ertain pages (i know it's not connected but still alout SSR, there was an issue for this0. This would prevent a lot of SSR errors that we have right now in pages that just don't need to use it.

**Comment by pkarw** [comment]
Filip its done this way - with one component which is switchable inside - App.vue? I didn’t get this point?

**Comment by pkarw** [comment]
We can theoretically filter the urls from SSR routing - however I don’t see many use cases for that. How wupuld you like to use it?

**Comment by filrak** [comment]
basically it's like this:
App.vue has 'layout' data property and dynamic component that renders appropiate compoennt based on value of `layout` data prop. Also we just HAVE 'layouts' folder with layouts to render in this dynamic component slot. That's all, few lines of code'

Abstract of the solution:
```<template>
  <component :is="layout" />
</template>
import { LayoutA } from './layouts/A.vue'
import { LayoutA } from './layouts/B.vue'

data: {
  return {
    layout: layoutA // will be a  computed property taking layout to render from vuex so it can be set in vuex in the app dynamically
  }
}

**Comment by filrak** [comment]
@pkarw i added 2 things to the post btw, about modules and ssr plugin vue

**Comment by filrak** [comment]
We can theoretically filter the urls from SSR routing - however I don’t see many use cases for that. How wupuld you like to use it?

We discussed it earlier. For SEO ssr is useful only for homepage and catalog. By disabling ti on unnecesary pages people will be able to use client-side libs without any problems which will be a huge benefit in dev experience

File: core/build/webpack.base.config.js [hunk 3]
```diff
@@ -40,7 +44,19 @@ const postcssConfig =  {
 module.exports = {
   plugins: [
     new CaseSensitivePathsPlugin(),
-    new VueLoaderPlugin()
+    new VueLoaderPlugin(),
+    // generate output HTML
```
Comment (filrak) [inline core/build/webpack.base.config.js]: maybe this files should be provided in config or exported from some place that develoeprs can reach (not from core)?

**Pull request merged**